### PR TITLE
PR#7331: ocamldoc, avoid self-alias induced infinite loop

### PR DESCRIPTION
**Pull request by Octachron**
[MPR#7331](https://caml.inria.fr/mantis/view.php?id=7331):
This PR fixes an infinite loop within ocamldoc in presence of alias loops in module definitions:
```(* a.ml *)
module rec A: sig 
(** Aphasia *)
type t end
= B
and B:sig type t= A.t end = A
```
Note that the fix is partial: more complex cycles may still trigger infinite loops. Moreover, the generated documentation is useless, since ocamldoc ignores documentation comments in signature constraints (i.e. `Aphasia` is not picked up by ocamldoc).

However, since this infinite loop is triggered for valid OCaml code, I would argue that this partial fix should go into 4.05, without waiting for a more complete (and complex) solution.

### CODE REVIEW

File: testsuite/tests/tool-ocamldoc-html/Makefile [hunk 1]
```diff
@@ -33,10 +33,11 @@ default:
 	fi
 
 .PHONY: run
-run: *.mli
-	@for file in *.mli; do \
+run: *.mli *.ml
+	@for file in *.ml *.mli; do \
 	  printf " ... testing '$$file'"; \
 	  F="`basename $$file .mli`"; \
+	  F="`basename $$F .ml`"; \
```
Comment (gasche) [inline testsuite/tests/tool-ocamldoc-html/Makefile]: Are you overwriting the previous `F` variable here? I'm a bit confused, are the `.mli` still tested?

**Comment by gasche** [comment]
I'm not completely convinced, but it may be because you are not selling your work very well.

I thought that we knew how to detect cycles in general, by keeping track of the nodes already visited in a traversal. I see that there are bits and pieces of exactly this being done in the PR, so why does this solution not work? Is it because of a risk of confusing a cycle with the existence of unrelated submodules of the same name? (That is, could your cycle-detection be confused by name conflicts?). If so, is there a risk of regression for existing programs where ocamldoc did not loop before?

There is another easy approach to avoid looping: increment a counter and bail out (with a warning) after N steps. It is ugly, but also very simple. If your approach is both complex and incomplete, we may be getting the worse of both worlds. If we don't actually know how to detect all cycles, what's wrong with the counter solution?

File: testsuite/tests/tool-ocamldoc-html/Makefile (hunk 1 above)
Comment (Octachron) [reply testsuite/tests/tool-ocamldoc-html/Makefile]: The mli are still tested: I am updating `F`  by removing `.ml` extensions after having removed the `.mli` ones: as far as I can see on its man page, `basename` can only removes one suffix at a time.

**Comment by Octachron** [comment]
> why does this solution not work?

Let me rephrase my thougts, all cycles in the digraph of module aliasing are detected, i.e cycles due to recursive definition of the form `module A_1 = A_2 … module A_n = A_1` . I am however not asserting that it is impossible to introduce more indirect cycles in recursive module definitions and make ocamldoc loop (after all even the type-checker have infinite loops with recursive modules).

> If so, is there a risk of regression for existing programs where ocamldoc did not loop before?

No, the new code only tracks visited module aliases. Since ocamldoc only uses module names at this point, any cycle detected with the code in this PR would have sent ocamldoc looping.

**Comment by gasche** [comment]
Thanks, this is better indeed that I was led to believe by your first message. But if you have this general solution that works, why do you also have tests like `if m.m_name <> ma.ma_name then` that look like they are just trying to prevent self-cycles?

**Comment by Octachron** [comment]
Rereading these changes in `odoc_cross.ml`, they were indeed a bad idea: they were more of a logical optimisation: do not add self aliases as aliases to be resolved in the future. 

The fact that this small optimisation is not followed by any cycle detection logic stems from the fact that the alias related code in `odoc_cross.ml` is incomplete and does not go anywhere in its current state. It just fills an unused hashtable (`modules_aliases` which  is unreachable from outside of `odoc_cross.ml`).

File: ocamldoc/odoc_module.ml [hunk 2]
```diff
@@ -219,37 +219,47 @@ let included_modules l =
...
+
 (** Returns the list of elements of a module.
-   @param trans indicates if, for aliased modules, we must perform a transitive search.*)
-let rec module_elements ?(trans=true) m =
+   @param trans indicates if, for aliased modules, we must perform a transitive search.
+   @param visited is used to guard against aliases loop
+     (e.g [module rec M:sig end=M] induced loop.
+
+**)
+let rec module_elements ?(visited=S.empty) ?(trans=true) m =
```
Comment (gasche) [inline ocamldoc/odoc_module.ml]: Nitpicking, but I don't like the use of an optional argument here. Not passing the argument is quite likely to be a programming error, so it should not be optional. Besides, you do rebind a more convenient definition (with empty visited set) later.

File: ocamldoc/odoc_module.ml [hunk 3]
```diff
@@ -459,20 +474,22 @@ let rec module_type_is_functor mt =
...
-  let rec iter = function
+  let rec iter visited = function
       Module_functor _ -> true
     | Module_alias ma ->
         (
-         match ma.ma_module with
-           None -> false
-         | Some (Mod mo) -> iter mo.m_kind
-         | Some (Modtype mt) -> module_type_is_functor mt
+          not (ma.ma_name = m.m_name || S.mem ma.ma_name visited)
```
Comment (gasche) [inline ocamldoc/odoc_module.ml]: Again, wouldn't the one-step condition be redundant here?

File: ocamldoc/odoc_module.ml (hunk 2 above)
Comment (Octachron) [reply ocamldoc/odoc_module.ml]: True.

File: ocamldoc/odoc_module.ml (hunk 3 above)
Comment (Octachron) [reply ocamldoc/odoc_module.ml]: Partially, it can be rewritten more cohesively however.

File: testsuite/tests/tool-ocamldoc-html/Makefile (hunk 1 above)
Comment (gasche) [reply testsuite/tests/tool-ocamldoc-html/Makefile]: So in bash `${F%.*}` works to strip the extension, but I'm fine with this instead which might be more portable (?). Maybe having a comment (`# stripping both .ml and .mli extensions`) could help people less familiar with this use of `basename`.

Comment (Octachron) [reply testsuite/tests/tool-ocamldoc-html/Makefile]: I am also not sure about portability, so I went with the comment option.

**Comment by gasche** [comment]
Thanks, merged.

**Comment by gasche** [comment]
(I should have squashed, but whatever.)

Cherry-picked in trunk ( 13205fca1d04ecfda32b8670643b228aa8922516 ).
